dags/prio/kubernetes.py

```diff
@@ -0,0 +1,109 @@
+    server_id,
+    env_vars,
+    arguments,
+    machine_type="n1-standard-1",
+    image="mozilla/prio-processor:latest",
+    location="us-west1-b",
+    owner_label="hwoo",
+    team_label="dataops",
+):
+    """Run a command on an ephemeral container running the
+    `mozilla/prio-processor:latest` image.
+
+    :param str parent_dag_name:         Name of the parent DAG.
+    :param str child_dag_name:          Name of the child DAG.
+    :param Dict[str, Any] default_args: Default arguments for the child DAG.
+    :param str gcp_conn_id:             Name of the connection string.
+    :param str service_account:         The address of the service account.
+    :param str server_id:               The identifier for the Prio processor
+    :param Dict[str, str] env_vars:     Environment variables for configuring
+                                        the processor.
+    :param List[str] arguments:         The command to run after loading the
+                                        image.
+    :param str machine_type:            The machine type for running the image.
+    :param str image:                   Dockerhub image
+    :param str location:                The region of the GKE cluster.
+    :param str owner_label:             Label for associating the owner
+    :param str team_label:              Label for associating the team
+    :return: DAG
+    """
+    assert server_id in ["a", "b", "admin"]
+
+    connection = GoogleCloudBaseHook(gcp_conn_id=gcp_conn_id)
+
+    cluster_name = "gke-prio-{}".format(server_id)
+
+    shared_config = {
+        "project_id": connection.project_id,
+        "gcp_conn_id": gcp_conn_id,
+        "location": location,
+    }
+
+    with DAG(
+        "{}.{}".format(parent_dag_name, child_dag_name), default_args=default_args
+    ) as dag:
+        create_gke_cluster = GKEClusterCreateOperator(
+            task_id="create_gke_cluster",
+            body=create_gke_config(
+                name=cluster_name,
+                service_account=service_account,
+                owner_label=owner_label,
+                team_label=team_label,
+                machine_type=machine_type,
+                # DataProc clusters require VPC with auto-created subnets
+                subnetwork="default" if server_id == "admin" else "gke-subnet",
+                is_dev=environ.get("DEPLOY_ENVIRONMENT") == "dev",
+            ),
+            **shared_config
+        )
+
+        # Running the pod without any time in-between will cause the scope-based
+        # authentication in Google Cloud Platform to fail. For example:
+        #
+        # `ServiceException: 401 Anonymous caller does not have
+        # storage.objects.get access to moz-fx-prio-dev-a-private/processed/`
+        #
+        # Sleeping by a small amount solves this problem. This issue was first
+        # noticed intermittently on 2019-09-09.
+        sleep = BashOperator(task_id="sleep", bash_command="sleep 15", dag=dag)
```

Comment:
How was 15 seconds determined? This seems like it may be an unreliable workaround, as opposed to having the container poll/retry. Another option would be pod readiness probes but it doesn't look like they are supported in k8spod or gkepod operators.

Reply:
I tried setting `retry=3` with `retry_delay=timedelta(seconds=30)`, but I can't seem to make the subdag task to retry. I did noticed that the 15 second timeout is sometimes not long enough, so I've bumped the timeout to 60 seconds. It's frustrating that the GKE cluster still needs more time to initialize the authorization flow.

I think the next round of deployment should run directly on top of GCE instances with the service account being generated in the instance and mounted into a docker instance. I think this would also help with the 5-7 minutes of GKE startup and also being able to run this on another provider.

Reply:
Oh i meant having the prio container retry until authentication is set up

Reply:
I'm not sure this would solve it, I think it might be at the cluster level as opposed to the container. I'll look into reproducing this in the `prio-processor` side using GCE.